src/perf_tool_backend.ml

```diff
@@ -113,58 +113,90 @@ module Perf_line = struct
-          match Re.Group.all (Re.exec kind_and_ip_re kind_and_ip) with
-          | [| _; kind; ip_string; ip_rest |] ->
-            let ip =
-              try Scanf.sscanf ip_string "%Lx" (fun ip -> ip) with
-              | Scanf.Scan_failure _ | End_of_file -> 0L
-            in
-            let parse_symbol_offset str =
-              try Scanf.sscanf str "%s@+0x%x" (fun symbol offset -> symbol, offset) with
-              | Scanf.Scan_failure _ | End_of_file -> "[unknown]", 0
-            in
-            let ip_symbol, ip_offset = parse_symbol_offset ip_rest in
-            let symbol, offset = parse_symbol_offset rest in
-            { Event.thread =
-                { pid = (if pid = 0 then None else Some (Pid.of_int pid))
-                ; tid = (if tid = 0 then None else Some tid)
-                }
-            ; time = time_lo + (time_hi * 1_000_000_000) |> Time_ns.Span.of_int_ns
-            ; kind =
-                (match String.strip kind with
-                | "call" -> Call
-                | "return" -> Return
-                | "tr strt" -> Start
-                | "tr end" -> End None
-                | "tr end  call" -> End Call
-                | "tr end  return" -> End Return
-                | "tr end  syscall" -> End Syscall
-                | "jmp" -> Jump
-                | "jcc" -> Jump
-                | kind -> raise_s [%message "unrecognized perf event" (kind : string)])
-            ; addr
-            ; symbol
-            ; offset
-            ; ip
-            ; ip_symbol
-            ; ip_offset
+      match Re.Group.all (Re.exec re line) with
+      | [| _
+         ; pid
+         ; tid
+         ; time_hi
+         ; time_lo
+         ; kind
+         ; src_instruction_pointer
+         ; src_symbol_and_offset
+         ; dst_instruction_pointer
+         ; dst_symbol_and_offset
+        |] ->
+        let pid = Int.of_string pid in
+        let tid = Int.of_string tid in
+        let time_lo =
+          (* In practice, [time_lo] seems to always be 9 decimal places, but it seems good to guard
+             against other possibilities. *)
+          let num_decimal_places = String.length time_lo in
+          match Ordering.of_int (Int.compare num_decimal_places 9) with
+          | Less -> Int.of_string time_lo * Int.pow 10 (9 - num_decimal_places)
+          | Equal -> Int.of_string time_lo
+          | Greater -> Int.of_string (String.prefix time_lo 9)
+        in
+        let time_hi = Int.of_string time_hi in
+        let int64_of_hex_string str =
+          try Scanf.sscanf str "%Lx" Fn.id with
+          | Scanf.Scan_failure _ | End_of_file -> 0L
+        in
+        let src_instruction_pointer = int64_of_hex_string src_instruction_pointer in
+        let dst_instruction_pointer = int64_of_hex_string dst_instruction_pointer in
+        let parse_symbol_and_offset str =
+          try Scanf.sscanf str "%s@+0x%x" (fun symbol offset -> symbol, offset) with
```

Comment:
Would you expect us to see any perf improvements by using a pre-baked regex here, over re-parsing this scan string twice per line? Not necessary to change in this PR, but I'm thinking it might be worth doing -- decoding the trace takes a long time right now.

Reply:
On reflection I realized that regardless of the performance issue, this is also a bug. We'll return "[unknown]" on symbols containing "+".  Changing to a regex fixes it, so I added a test and did that. Not sure how or whether it will affect the performance because I don't have a good performance model of either Scanf or Re.

Reply:
Good catch!